engines/lab/interface.cpp

```diff
@@ -0,0 +1,154 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+/*
+ * This code is based on Labyrinth of Time code with assistance of
+ *
+ * Copyright (c) 1993 Terra Nova Development
+ * Copyright (c) 2004 The Wyrmkeep Entertainment Co.
+ *
+ */
+
+#include "common/events.h"
+
+#include "lab/lab.h"
+
+#include "lab/dispman.h"
+#include "lab/eventman.h"
+#include "lab/image.h"
+#include "lab/utils.h"
+
+namespace Lab {
+
+Button *EventManager::createButton(uint16 x, uint16 y, uint16 id, uint16 key, Image *image, Image *altImage) {
+	Button *button = new Button();
+
+	if (button) {
+		button->_x = _vm->_utils->vgaScaleX(x);
+		button->_y = y;
+		button->_buttonId = id;
+		button->_keyEquiv = key;
+		button->_image = image;
+		button->_altImage = altImage;
+		button->_isEnabled = true;
+
+		return button;
+	} else
+		return nullptr;
+}
+
+void EventManager::freeButtonList(ButtonList *buttonList) {
+	for (ButtonList::iterator buttonIter = buttonList->begin(); buttonIter != buttonList->end(); ++buttonIter) {
+		Button *button = *buttonIter;
+		delete button->_image;
+		delete button->_altImage;
+		delete button;
+	}
+
+	buttonList->clear();
+}
+
+void EventManager::drawButtonList(ButtonList *buttonList) {
+	for (ButtonList::iterator button = buttonList->begin(); button != buttonList->end(); ++button) {
+		toggleButton((*button), 1, true);
+
+		if (!(*button)->_isEnabled)
+			toggleButton((*button), 1, false);
+	}
+}
+
+void EventManager::toggleButton(Button *button, uint16 disabledPenColor, bool enable) {
+	if (!enable)
+		_vm->_graphics->overlayRect(disabledPenColor, button->_x, button->_y, button->_x + button->_image->_width - 1, button->_y + button->_image->_height - 1);
+	else
+		button->_image->drawImage(button->_x, button->_y);
+
+	button->_isEnabled = enable;
+}
+
+uint16 EventManager::makeButtonKeyEquiv(uint16 key) {
+	if (Common::isAlnum(key))
+		key = tolower(key);
+
+	return key;
+}
+
+Button *EventManager::checkNumButtonHit(ButtonList *buttonList, uint16 key) {
+	uint16 gkey = key - '0';
```

Comment:
`key` seems to come from a `Common::KeyCode` (inside `EventManager::getMsg`). Doing arithmetic like it is an ASCII character is something you should avoid at any cost.
Overall, I suggest to use proper types to avoid these issues from coming up. Here's a short list of suggestions:
1) Make `createButton` take a `Common::KeyCode` for `key`, adjust `Button`'s `_keyEquiv` to be a `Common::KeyCode` too, and finally make all callers of `createButton` pass key codes instead of ASCII values or similar (which is the case right now).
2) I think this allows to get rid of `makeButtonKeyEquiv` too. Right now it treats its input as ASCII even though it's always a `Common::KeyCode`. Thus, the `tolower` etc. is unnecessary because key codes work different from ASCII codes.
3) Make this function take a `Common::KeyCode` for `key` and adapt it properly, i.e. the `gkey` type etc. too.
4) Adapt all code using `VKEY_UPARROW` etc. to use `Common::KeyCode` values too.
5) Make `IntuiMessage::_code` a `Common::KeyCode` too to reflect that it stores values of that type. This might not be trivial because it seems there is an inflow from a `Button`'s `_buttonId` too. Thus, some more tickeling down or other efforts to increase (type) safety might be required.